pkg/server/sotw/v2/server.go

```diff
@@ -195,78 +168,13 @@ func (s *server) process(stream Stream, reqCh <-chan *discovery.DiscoveryRequest
 		case <-s.ctx.Done():
 			return nil
 		// config watcher can send the requested resources types in any order
-		case resp, more := <-values.endpoints:
-			if !more {
-				return status.Errorf(codes.Unavailable, "endpoints watch failed")
-			}
-			nonce, err := send(resp, resource.EndpointType)
-			if err != nil {
-				return err
-			}
-			values.endpointNonce = nonce
-
-		case resp, more := <-values.clusters:
-			if !more {
-				return status.Errorf(codes.Unavailable, "clusters watch failed")
-			}
-			nonce, err := send(resp, resource.ClusterType)
+		case resp := <-values.responses:
+			typeUrl := resp.GetRequest().TypeUrl
+			nonce, err := send(resp, typeUrl)
```

Comment:
The way would be to return an error immediately, or as @fxposter mentions, we could send a token response as an erroneous value.

Reply:
sounds good...a killpill to represent closing channels sounds good, but it will involve much more code and maintenance for the control plane to do it. Right now, its easy to close channels. May be we can show how to perform that as part of the PR?

Reply:
How closing channel is different than sending nil to it?

Reply:
yeah..sending nil is ok. I was not sure what the special token would be.

Reply:
although, im not sure if sending nil is ok from a xds protocol stand point. The special killpill should be special enough and not a nil, in case nil is a valid xds response. In our private control plane, we dont return nil today, but there is no guardrails to say nil is a special value.

Reply:
may be another way to deal with this could be passing a cancel func , and the control plane implementations can call cancelFunc to denote closed channel

Reply:
https://github.com/envoyproxy/go-control-plane/blob/master/pkg/server/sotw/v2/server.go#L165-L168 - current implementation already handles `nil` as an error and propagates it further.

Reply:
i see.. using nil as killpill sounds good.